collector/mdadm.go

```diff
@@ -0,0 +1,200 @@
+
+	sizeDirty := strings.Split(first_split[0], "blocks")[0]
+	sizeStripped := strings.Replace(sizeDirty, " ", "", -1)
+
+	numberOfDrives := strings.Split(second_split[0], "/")
+
+	total, _ = strconv.ParseInt(numberOfDrives[0], 10, 64)
+	active, _ = strconv.ParseInt(numberOfDrives[1], 10, 64)
+	size, _ = strconv.ParseInt(sizeStripped, 10, 64)
+
+	return active, total, size
+
+}
+
+// gets the size that has already been synced out of the sync-line
+func evalBuildline(buildline string) int64 {
+	first_split := strings.Split(buildline, "(")
+	size := strings.Split(first_split[1], "/")[0]
+
+	syncedSize, _ := strconv.ParseInt(size, 10, 64)
+
+	return syncedSize
+}
+
+// parses an mdstat-file and returns a struct with the relevant infos
+func parseMdstat(mdstatusfilepath string) ([]mdstatus, error) {
+
+	content, err := ioutil.ReadFile(mdstatusfilepath)
+	if err != nil {
+		return []mdstatus{}, err
+	}
+
+	mdstatusfile := string(content)
+
+	lines := strings.Split(mdstatusfile, "\n")
+	var currentMd string
+
+	// each md has at least the deviceline, statusline and one empty line afterwards
+	// so we will have probably something of the order len(lines)/3 devices
+	// so we use that for preallocation
+	estimateMds := len(lines) / 3
+	mdstates := make([]mdstatus, 0, estimateMds)
+
+	for i, l := range lines {
+		if !(len(l) > 0) {
+			// skip entirely empty lines
+			continue
+		}
+
+		if string(l[0]) == " " {
+			// those lines are not the beginning of a md-section
+			continue
+		} else {
+
+			if strings.HasPrefix(l, "Personalities") || strings.HasPrefix(l, "unused") {
+				// we aren't interested in lines with general info
+				continue
+			}
+
+			currentMd = strings.Split(l, " ")[0]         // name of the md-device we are parsing atm
+			isActive := !strings.Contains(l, "inactive") // activity-status of said md-device
+
+			active, total, size := evalStatusline(lines[i+1]) // parse statusline, always present
+
+			// now get the number of synced blocks
+			var syncedBlocks int64
+
+			// get the line number of the syncing-line
+			var j int
+			if strings.Contains(lines[i+2], "bitmap") { // then skip the bitmap line
+				j = i + 3
+			} else {
+				j = i + 2
+			}
+
+			// if device is syncing atm, get the number of currently synced bytes,
+			// otherwise that number equals the size of the device
+			if strings.Contains(lines[j], "recovery") || strings.Contains(lines[j], "resync") {
+				syncedBlocks = evalBuildline(lines[j])
+			} else {
+				syncedBlocks = size
+			}
+
+			mdstates = append(mdstates, mdstatus{currentMd, isActive, active, total, size, syncedBlocks})
+
+		}
+
+	}
+
+	return mdstates, nil
+
+}
+
+// just returns th epointer to a struct holding all exportable metrics
+func NewMdadmCollector() (Collector, error) {
+
+	return &mdadmCollector{
+		isActive: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: "md_is_active",
+				Help: "indicator whether the md-device is active or not",
```

Comment:
To be consistent with other collectors and Prometheus components, please start metric help strings with a capital letter and end them with a period.